09_simulations/nbconverted/simulate_domains.py

```diff
@@ -0,0 +1,372 @@
+    simulate_csd_corr,
+)
+from models import (
+    train_ridge,
+    train_rf,
+    train_mlp,
+    get_prob_metrics
+)
+
+np.random.seed(42)
+
+get_ipython().run_line_magic('load_ext', 'autoreload')
+get_ipython().run_line_magic('autoreload', '2')
+
+
+# ### Simulate data from multiple "domains"
+# 
+# Main simulation code is in `09_simulations/csd_simulations.py`.
+
+# In[2]:
+
+
+n_domains = 5
+n_per_domain = 50
+p = 20
+k = 5
+noise_scale = 1.5
+
+simulate_with_csd = True
+simulate_same_z = True
+correlated_noise = True
+
+
+# In[3]:
+
+
+if k is not None:
+    if simulate_with_csd:
+        if correlated_noise:
+            xs, ys = simulate_csd_corr(n_domains, n_per_domain, p, k, corr_top=0.8, diag=1)
+        else:
+            xs, ys = simulate_csd(n_domains, n_per_domain, p, k, noise_scale)
+    else:
+        xs, ys = simulate_no_csd_large_z(n_domains, n_per_domain, p, k, noise_scale)
+elif simulate_same_z:
+    xs, ys = simulate_no_csd_same_z(n_domains, n_per_domain, p, noise_scale)
+else:
+    xs, ys = simulate_no_csd(n_domains, n_per_domain, p, noise_scale)
+
+print(xs.shape)
+print(xs[:5, :5])
+
+
+# In[4]:
+
+
+print(ys.shape)
+print(ys[:3, :])
+
+
+# ### Plot simulated data
+# 
+# We'll do this using both PCA and UMAP, side by side. We can color by domain and use different shapes for each label, to get an idea of how data clusters with respect to domain and how separable we expect different labels to be across domains.
+
+# In[5]:
+
+
+pca = PCA(n_components=2)
+X_proj_pca = pca.fit_transform(xs)
+reducer = UMAP(n_components=2, random_state=42)
+X_proj_umap = reducer.fit_transform(xs)
+
+domains = np.concatenate([([i] * n_per_domain) for i in range(n_domains)])
+
+X_pca_df = pd.DataFrame(X_proj_pca,
+                        columns=['PC{}'.format(j) for j in range(X_proj_pca.shape[1])])
+X_pca_df['domain'] = domains
+X_pca_df['label'] = ys.flatten()
+
+X_umap_df = pd.DataFrame(X_proj_umap,
+                        columns=['UMAP{}'.format(j) for j in range(X_proj_umap.shape[1])])
+X_umap_df['domain'] = domains
+X_umap_df['label'] = ys.flatten()
+
+X_umap_df.head()
+
+
+# In[6]:
+
+
+sns.set({'figure.figsize': (18, 8)})
+fig, axarr = plt.subplots(1, 2)
+
+sns.scatterplot(data=X_pca_df, x='PC0', y='PC1', hue='domain', style='label', s=50, ax=axarr[0])
+sns.scatterplot(data=X_umap_df, x='UMAP0', y='UMAP1', hue='domain', style='label', s=50, ax=axarr[1])
+    
+axarr[0].set_title('PCA projection of simulated data, colored by domain')
+axarr[1].set_xlabel('PC1')
+axarr[1].set_ylabel('PC2')
+axarr[0].legend()
+axarr[1].set_title('UMAP projection of simulated data, colored by domain')
+axarr[1].set_xlabel('UMAP dimension 1')
+axarr[1].set_ylabel('UMAP dimension 2')
+axarr[1].legend()
+
+
+# ### Random train/test splitting
+# 
+# Just split the data randomly here, and fit some models to the split data. This gives us an idea of what baseline performance we should expect when we have access to training data from all domains.
```

Comment:
For clarity in the first sentence, can you say explicitly that you're splitting data from all of the 4 domains?